fix #1012, add DataSource and DataSink API

manuzhang opened the pull request:
The API is a bit different from the design in #1012.

The major difference is that `DataSource` and `DataSink` are added as streaming interfaces rather than extending `Task`.
`KafkaSource` extends `TimeReplayableSource` which is a `DataSource` and `KafkaSink` extends `DataSink` respectively.

They are used in `KafkaSourceTask` and `KafkaSinkTask` which are provided as standard kafka connectors at task-level. 

`KafkaSourceProcessor` and `KafkaSinkProcessor` are utility objects to ease user's life of DAG setup and configuration.

Also, I've removed the Kafka RollingTopWords examples which lack maintenance and I'd like to implement with state api in the future.

whjiang commented on streaming/src/main/scala/org/apache/gearpump/streaming/source/DataSource.scala:
```diff
@@ -0,0 +1,53 @@
+   */
+  def read(num: Int): List[Message]
+
+  def close(): Unit
```
who will call this close() and read() and under which situation? Need better document?

whjiang commented on streaming/src/main/scala/org/apache/gearpump/streaming/transaction/api/TimeReplayableSource.scala:
```diff
@@ -40,24 +41,16 @@ import org.apache.gearpump.{Message, TimeStamp}
  *
  */
-trait TimeReplayableSource extends java.io.Serializable {
+trait TimeReplayableSource extends DataSource {
```
Better document, hard to understand these 2 functions. 

If time starts from 0, and if I setStartTime(10) and then call startFromBeginning(), what data will I read, from 0 or from 10?

whjiang commented:
Hi Tianlun,

I will prefer following refinements:
1. It is better if you can provide some general SourceTask and SinkTask which accepts a DataSource and a DataSink as arguments.Or a ReplayableSourceTask accepts ReplayableSourceTask. What I prefer is to make user pipe creation as simple as possible. E.g. you mentioned that user can insert a timestamp filter to perform such filtering. My question is: can we standardize it in ReplayableSourceTask? Or make the common used pattern standardized? 
2. Please remove KafkaConfig and use similar parameters as Spark Streaming does. My understanding is that KafkaConfig is our invented config format which is not friendly to users.

manuzhang commented:
Hi Weihua,

I've updated the PR to address your comments. 

1. Users only need to extend `DataSource` and `DataSink`, which would be run by the general `DataSourceTask` and `DataSinkTask`. The source/sink is created when users set up a DAG and passed to the general `DataSourceProcessor` and `DataSinkProcessor` (please check out the `KafkaWordCount` example). 

2. Why is the `KafkaConfig` unfriendly to users ? Kafka has a long list of configs that are very tedious to be passed in as parameters. We also have our own configs required for `KafkaSource`. Since we've already implemented the `KafkaSource` for users, they only need to care about `kafka.conf`. SparkStreaming also requires users to pass in `kafkaParams` through a map (see `KafkaUtils.createDirectStream` when it uses low level kafka api.

whjiang commented on experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/StreamApp.scala:
```diff
@@ -100,15 +100,13 @@ class TimeReplayableProducer[T:ClassTag](timeReplayableSource: TimeReplayableSou
-    kafkaSource.startFromBeginning
+  lazy val source = Some(new KafkaSource(kafkaConfig)).map(kafkaSource => {
+    // WARN: the kafka source doesn't work without calling open
     kafkaSource
```
I don't see the open() call. Is it a mistake?

manuzhang replied:
`open` requires `TaskContext` to be passed in which is not possible here.

manuzhang replied:
DSL currently doesn't work well the existing low-level api so I add a note here and we should address this as follow-ups

whjiang commented on external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/KafkaDecoder.scala:
```diff
@@ -18,17 +18,11 @@
 
-import scala.util.{Failure, Success}
-
 class KafkaDecoder extends MessageDecoder {
```
Better name as DefaultMessageDecoder as it is not specific to Kafka.

whjiang commented on external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/sink/KafkaSink.scala:
```diff
@@ -0,0 +1,49 @@
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
+import org.apache.kafka.common.serialization.ByteArraySerializer
+
+class KafkaSink(getProducer: () => KafkaProducer[Array[Byte], Array[Byte]], topic: String) extends DataSink {
```
Need better doc here. A little bit hard to understand the logic here.

whjiang commented on streaming/src/main/scala/org/apache/gearpump/streaming/source/DataSource.scala:
```diff
@@ -0,0 +1,66 @@
+   * invoked in each onNext() method of [[org.apache.gearpump.streaming.task.Task]]
+   * @return a list of messages wrapped in [[Message]]
+   */
+  def read(): List[Message]
```
what is the maximum batch size?

manuzhang replied:
which maximum batch size ? `DataSource` has no notion of `batch size`. That is specific to `KafkaSource` which is the number of messages to read and send downstreams in one `onNext`

whjiang replied:
read() return a list of messages. How to control the max size of the list?

manuzhang replied:
that is left to api developers.

whjiang commented on experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/StreamApp.scala:
```diff
@@ -100,15 +100,13 @@ class TimeReplayableProducer[T:ClassTag](timeReplayableSource: TimeReplayableSou
 }
 
 class KafkaProducer[T:ClassTag](kafkaConfig: KafkaConfig, converter: Message => T) extends Traverse[T] {
```
better move the KafkaProducer to kafka package instead of DSL package. Otherwise DSL package compilation will need kafka package.

manuzhang replied:
The changes in this PR to DSL or pipeline modules are just to make them compilable with the new API. I'll open follow-up issues to address them

whjiang commented on streaming/src/main/scala/org/apache/gearpump/streaming/source/DataSource.scala:
```diff
@@ -0,0 +1,66 @@
+ *
+ * subclass is required to be serializable
+ */
+trait DataSource extends java.io.Serializable {
```
I will prefer DataSource[T] instead of DataSource. T is the user message type.

manuzhang replied:
currently, we do not expose the type of message in the low-level api so I don't see the benefit of `DataSource[T]` here.

whjiang replied:
I will still prefer to expose T as this may benefit the high level API and future low level API. Remember we have a proposal to make Message type invisible to user.

manuzhang replied:
yes. I remember that so I'd like to defer the change till that comes in. Also, the high level API has its own `SourceTask`, which could (should ?) be merged with this source API in the future. The problem is we have so many flying pieces so I prefer to stick to the old way until we have a consistent to way to express (low-level api, high-level api, message type) throughout the system.

whjiang replied:
Ok with your idea.

kkasravi replied:
T would actually be Message[T] however this is a very difficult change having tried it several time and affects many files. See https://github.com/intel-hadoop/gearpump/issues/57 for some context.

kkasravi replied:
@manuzhang  SourceTask, SinkTask, SinkConsumer, TraverseType and Traverse types have a number of similarities and I do think these types as well as TraversableSource and TraversableSink should be merged into this API's. Right now using the high level API it's possible to never expose Message which I think is a good objective since Message is an internal transport wrapper. Also SourceTask is compatible with scala collections which is another nice objective.

manuzhang replied:
Yes. I totally agree. The only problem is that many are experimental and we can't resolve all the issues in one shot.

whjiang commented:
Hi Tianlun,

I am almost comfortable with the newest code except KafkaConfig. 

KafkaConfig is our invented config format for kafka. It mixed kafka config and our specific requirements like Decoder, timestamp etc. And also it has limitation under case like I need to read from two different kafka systems with different configurations, etc. 

My suggestion is:
1. Remove KafkaConfig from our API. Use the parameter list similar to Spark Streaming kafka connector.
2. You can keep KafkaConfig as a util class to help load/save such long parameter lists for user if you like. But, just a util class for user convenience. User is totally possible to bypass it.

manuzhang commented:
Hi Weihua,

As I said, the Spark Streaming kafka connector that allows a parameter list to be passed in uses high level kafka consumer API rather than the low level kafka consumer API used by our `KafkaSource`. Besides, `KafkaSource` requires both kafka consumer and producer configs to be passed in since `KafkaSource` also stores the mapping of offset to timestamp to kafka. Plus, some configs will be passed in a long code path. (`KafkaSource` -> `FetchThread` -> `KafkaConsumer`) .

> And also it has limitation under case like I need to read from two different kafka systems with different configurations, etc.

we don't keep users from loading different kafka configurations as long as they are from different conf files like `kafka1.conf` and `kafka2.conf`.

I agree that forcing user to use `KafkaConfig` is not nice but I don't see how it could be easier for users the other way around with a few code changes.

manuzhang commented:
@kkasravi the API change would affect Pipeline and PipelineDSL. I only make them compile in this PR and they won't work if the PR is merged in. I'd like to defer the fix to follow-ups since this PR is already quite big.  Is that OK for you ?

clockfly commented:
@manuzhang 

I am not OK that any new commit break existing code. Please fix Pipeline with new updates.

manuzhang commented:
@kkasravi @clockfly 

The PipelineDSL is working as it was now.  The UT failure is unrelated.

kkasravi commented on streaming/src/main/scala/org/apache/gearpump/streaming/source/DataSourceProcessor.scala:
```diff
@@ -0,0 +1,33 @@
+    Processor[DataSourceTask](parallelism, description = description,
+      taskConf.withValue[DataSource](DataSourceTask.DATA_SOURCE, dataSource))
+  }
+}
```
Maybe this should be something like
```
object DataSourceProcessor {
  def apply[T<:DataSource,U<:DataSourceTask](...)

manuzhang replied:
Hi Kam.

1. We will serialize `DataSource` object into `UserConfig` and deserialize it back in a running  `DataSourceTask`  here.
2. We have a general `DataSourceTask` that accepts any `DataSource` so users don't need to write their own.

manuzhang commented:
Guys, thank you all for the good comments.
As this is a big PR and there are still concerns with the KafkaConfig, I'd like to split it into several PRs. I opened #1049 to add the source and sink API only. Closed this for now.
